src/libraries/Microsoft.PowerFx.Json/Functions/JsonFunctionImpl.cs

```diff
@@ -129,17 +132,56 @@ private JsonFlags GetFlags()
 
             private class Utf8JsonWriterVisitor : IValueVisitor
             {
+                private const int _maxDepth = 20;
+                private const int _maxLength = 1048576; // 1 MB
+
                 private readonly Utf8JsonWriter _writer;
                 private readonly TimeZoneInfo _timeZoneInfo;
                 private readonly bool _flattenValueTables;
+                private readonly Func<int, bool> _checkCancellation;
 
                 internal readonly List<ErrorValue> ErrorValues = new List<ErrorValue>();
 
-                internal Utf8JsonWriterVisitor(Utf8JsonWriter writer, TimeZoneInfo timeZoneInfo, bool flattenValueTables)
+                internal Utf8JsonWriterVisitor(Utf8JsonWriter writer, TimeZoneInfo timeZoneInfo, bool flattenValueTables, Action checkCancellation)
                 {
                     _writer = writer;
                     _timeZoneInfo = timeZoneInfo;
                     _flattenValueTables = flattenValueTables;
+
+                    _checkCancellation = (int index) =>
+                    {
+                        checkCancellation();
+
+                        bool error = false;
+
```

Comment:
We don't need to return bool. 

Just throw OperationCanceledException; 

Particularly because checkCancellation already throws , so there's no need to introduce a 2nd return convention (where cancel could either throw or return true)

Reply:
it might be simpler to just


1. save the _checkCancellation directly as a field, rather than the field be a newly created action that closes over the parameter. 

2. have an instance method:

private void CheckCancel(int depth) 
{
  this._checkCancellation(); // this throws 
  // other checks, which also throw

}